src/sys/socket/mod.rs

```diff
@@ -567,57 +569,95 @@ impl<'a> ControlMessage<'a> {
+            ControlMessage::Unknown(ref cmsg) => cmsg.0.cmsg_type,
+        }
+    }
+
     // Unsafe: start and end of buffer must be cmsg_align'd. Updates
     // the provided slice; panics if the buffer is too small.
     unsafe fn encode_into(&self, buf: &mut [u8]) {
-        match *self {
-            ControlMessage::ScmRights(fds) => {
-                let cmsg = cmsghdr {
-                    cmsg_len: self.len() as _,
-                    cmsg_level: libc::SOL_SOCKET,
-                    cmsg_type: libc::SCM_RIGHTS,
-                    ..mem::uninitialized()
-                };
-                let buf = copy_bytes(&cmsg, buf);
-
-                let padlen = cmsg_align(mem::size_of_val(&cmsg)) -
-                    mem::size_of_val(&cmsg);
-                let buf = pad_bytes(padlen, buf);
-
-                let buf = copy_bytes(fds, buf);
-
-                let padlen = self.space() - self.len();
-                pad_bytes(padlen, buf);
-            },
-            ControlMessage::ScmTimestamp(t) => {
-                let cmsg = cmsghdr {
-                    cmsg_len: self.len() as _,
-                    cmsg_level: libc::SOL_SOCKET,
-                    cmsg_type: libc::SCM_TIMESTAMP,
-                    ..mem::uninitialized()
-                };
-                let buf = copy_bytes(&cmsg, buf);
-
-                let padlen = cmsg_align(mem::size_of_val(&cmsg)) -
-                    mem::size_of_val(&cmsg);
-                let buf = pad_bytes(padlen, buf);
-
-                let buf = copy_bytes(t, buf);
-
-                let padlen = self.space() - self.len();
-                pad_bytes(padlen, buf);
-            },
-            ControlMessage::Unknown(UnknownCmsg(orig_cmsg, bytes)) => {
-                let buf = copy_bytes(orig_cmsg, buf);
+        let final_buf = if let ControlMessage::Unknown(ref cmsg) = *self {
+            let &UnknownCmsg(orig_cmsg, bytes) = cmsg;
 
-                let padlen = cmsg_align(mem::size_of_val(&orig_cmsg)) -
-                    mem::size_of_val(&orig_cmsg);
-                let buf = pad_bytes(padlen, buf);
+            let buf = copy_bytes(orig_cmsg, buf);
 
-                let buf = copy_bytes(bytes, buf);
+            let padlen = cmsg_align(mem::size_of_val(&orig_cmsg)) -
+                mem::size_of_val(&orig_cmsg);
+            let buf = pad_bytes(padlen, buf);
 
-                let padlen = self.space() - self.len();
-                pad_bytes(padlen, buf);
+            copy_bytes(bytes, buf)
+        } else {
+            let cmsg = cmsghdr {
+                cmsg_len: self.len() as _,
+                cmsg_level: libc::SOL_SOCKET,
+                cmsg_type: self.type_(),
+                ..mem::zeroed() // zero out platform-dependent padding fields
+            };
+            let buf = copy_bytes(&cmsg, buf);
+
+            match *self {
+                ControlMessage::ScmRights(fds) => {
+                    let padlen = cmsg_align(mem::size_of_val(&cmsg)) -
```

Comment:
It looks like `padlen` and buf are identical for all cases.  Can you move their definitions outside of the `match` block?